# 4] SVMs:


    SUBJECT: Foundations of Machine Learning(CS5590)

    NAME: Aviraj Antala  
    Roll No:CS24MTECH14011

In [13]:
import pandas as pd
import numpy as np
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

#First for apply svm we need to load the data 
#here sep='\s+' use to handle space or tabs between data header none means there is no header of column
train_data = pd.read_csv("http://www.amlbook.com/data/zip/features.train", sep='\s+', header=None)
test_data = pd.read_csv("http://www.amlbook.com/data/zip/features.test", sep='\s+', header=None)

# selecting the row where number is either 1 or 5
# here we make copy so we can avoid changes in original data 
train_with_0_5= train_data[(train_data[0] == 1) | (train_data[0] == 5)].copy()
test_with_0_5 = test_data[(test_data[0] == 1) | (test_data[0] == 5)].copy()

# in question it is given that consider 1 as a positive class and 5 as negative class 
train_with_0_5[0] = train_with_0_5[0].where(train_with_0_5[0] == 1, -1)
test_with_0_5[0] = test_with_0_5[0].where(test_with_0_5[0] == 1, -1)

# now we know that first column is target labels so we need to seperate lables and features
X_train = train_with_0_5.loc[:, 1:].values
y_train = train_with_0_5.iloc[:, 0].values

X_test = test_with_0_5 .iloc[:, 1:].values
y_test = test_with_0_5 .iloc[:, 0].values

# now we need to train the model so first create the model
# here we can use c as regularization parameter
linear_svm_model = SVC(kernel='linear')
linear_svm_model.fit(X_train, y_train)

# Predict and evaluate
y_pred = linear_svm_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
total_supportvector = len(linear_svm_model.support_vectors_)
print("Answer 4_a)\n")
print(f"Accuracy on test data: {accuracy:.2f}({accuracy*100:.2f}%)")
print(f"Total Number of Support Vectors: {total_supportvector}")


Answer 4_a)

Accuracy on test data: 0.98(97.88%)
Total Number of Support Vectors: 28


    Answer 4_a)

* Total accuracy for the test data=97.88%  
* Total number of support vector= 28  
* Also we know that support vector and accuracy depends on regularization parameter C.
* With C=2 then support vector is 26.
* With C=100 then accuracy is 98.11% that show that if C large then it give less misclassification but it leads to overfitting also.


    Answer 4_b)

In [26]:
# now we have to take different size as a input so for better we make a list of input
datasize = [50, 100, 200, 800]
print("Answer 4_b)\n")
for size in datasize:
    X_train_1_5 = X_train[:size] #this shows that for each loop we take train data upto 50, 100, 200, or 800
    y_train_1_5 = y_train[:size] #same for lables

    # train the svm with different size of data
    linear_svm_size= SVC(kernel='linear', C=1)
    linear_svm_size.fit(X_train_1_5, y_train_1_5)

    # predict value on test data
    y_pred_size = linear_svm_size.predict(X_test)
    accuracy_subset = accuracy_score(y_test, y_pred_size)
    total_supportvector_size = len(linear_svm_size.support_)

    
    print(f"Size of train data: {size}")
    print(f"Accuracy: {accuracy_subset:.4f}({accuracy_subset*100:.2f}%)")
    print(f"Total Number of Support Vectors: {total_supportvector_size}\n")


Answer 4_b)

Size of train data: 50
Accuracy: 0.9811(98.11%)
Total Number of Support Vectors: 2

Size of train data: 100
Accuracy: 0.9811(98.11%)
Total Number of Support Vectors: 4

Size of train data: 200
Accuracy: 0.9811(98.11%)
Total Number of Support Vectors: 8

Size of train data: 800
Accuracy: 0.9811(98.11%)
Total Number of Support Vectors: 14



In [27]:
# we have to check with Q=2 and Q=5
# C values is given in question 
Q_value = [2, 5]
C_value = [0.0001, 0.001, 0.01, 1]
print("Answer 4_c)\n")
for C in C_value:
    for Q in Q_value:
        poly_svm_model = SVC(kernel='poly', degree=Q, C=C) # now we have to make polynomial kernal
        poly_svm_model.fit(X_train, y_train)

        # training error
        y_train_pred_Q_C = poly_svm_model.predict(X_train)
        train_error_Q_C = (1 - accuracy_score(y_train, y_train_pred_Q_C))

        # total number of support vectors
        total_support_vectors_poly = len(poly_svm_model.support_)

        # test error
        y_test_pred_Q_C = poly_svm_model.predict(X_test)
        test_error_Q_C = (1 - accuracy_score(y_test, y_test_pred_Q_C))

        print(f"Q = {Q}, C = {C}")
        print(f"Training Error: {train_error_Q_C:.4f} ({train_error_Q_C * 100:.2f}%)")
        print(f"Test Error: {test_error_Q_C:.4f} ({test_error_Q_C * 100:.2f}%)")
        print(f"Total Number of Support Vectors: {total_support_vectors_poly}\n")



Answer 4_c)

Q = 2, C = 0.0001
Training Error: 0.3408 (34.08%)
Test Error: 0.3467 (34.67%)
Total Number of Support Vectors: 1112

Q = 5, C = 0.0001
Training Error: 0.0519 (5.19%)
Test Error: 0.0755 (7.55%)
Total Number of Support Vectors: 374

Q = 2, C = 0.001
Training Error: 0.0250 (2.50%)
Test Error: 0.0354 (3.54%)
Total Number of Support Vectors: 558

Q = 5, C = 0.001
Training Error: 0.0211 (2.11%)
Test Error: 0.0307 (3.07%)
Total Number of Support Vectors: 158

Q = 2, C = 0.01
Training Error: 0.0083 (0.83%)
Test Error: 0.0212 (2.12%)
Total Number of Support Vectors: 164

Q = 5, C = 0.01
Training Error: 0.0083 (0.83%)
Test Error: 0.0212 (2.12%)
Total Number of Support Vectors: 68

Q = 2, C = 1
Training Error: 0.0045 (0.45%)
Test Error: 0.0189 (1.89%)
Total Number of Support Vectors: 30

Q = 5, C = 1
Training Error: 0.0045 (0.45%)
Test Error: 0.0165 (1.65%)
Total Number of Support Vectors: 26



    Answer 4_c)

<i><b>i) When C = 0.0001, training error is higher at Q = 5.</i> </b> 
When Q=2 training error =34.08%  
When Q=5 training error =5.19%  
        Anwer: False  <br><br>
<i><b>ii) When C = 0.001, the number of support vectors is lower at Q = 5.</i> </b>   
When Q=2 number of support vector = 164  
When Q=5 number of support vector = 68  
        Anwer: True  <br><br>  
<i><b>iii) When C = 0.01, training error is higher at Q = 5. </i> </b>  
When Q=2 training error =0.83%    
When Q=5 training error =0.83%    
        Anwer: flase <br><br>
<i><b>iv) When C = 1, test error is lower at Q = 5</i> </b>   
When Q=2 testing error =1.89%  
When Q=5 testing error =1.65%  
        Anwer: True  <br><br>

In [28]:
# different c values
C_value_rbf = [0.01, 1, 100, 10000, 1000000]
print("Answer 4_d)\n")
for C in C_value_rbf:
    rbf_svm_model = SVC(kernel='rbf', C=C) #now we have to create rbf kernal
    rbf_svm_model.fit(X_train, y_train)

    # training error
    y_train_pred_rbf = rbf_svm_model.predict(X_train)
    train_error_rbf = (1 - accuracy_score(y_train, y_train_pred_rbf))

    # test error
    y_test_pred_rbf = rbf_svm_model.predict(X_test)
    test_error_rbf = (1 - accuracy_score(y_test, y_test_pred_rbf))
    total_support_vectors_rbf = len(rbf_svm_model.support_)

    print(f"C = {C}")
    print(f"Training Error: {train_error_rbf:.4f} ({train_error_rbf * 100:.2f}%)")
    print(f"Test Error: {test_error_rbf:.4f} ({test_error_rbf * 100:.2f}%)")
    print(f"Total Number of Support Vectors: {total_support_vectors_rbf}\n")



Answer 4_d)

C = 0.01
Training Error: 0.0051 (0.51%)
Test Error: 0.0165 (1.65%)
Total Number of Support Vectors: 345

C = 1
Training Error: 0.0045 (0.45%)
Test Error: 0.0212 (2.12%)
Total Number of Support Vectors: 33

C = 100
Training Error: 0.0032 (0.32%)
Test Error: 0.0189 (1.89%)
Total Number of Support Vectors: 24

C = 10000
Training Error: 0.0026 (0.26%)
Test Error: 0.0189 (1.89%)
Total Number of Support Vectors: 18

C = 1000000
Training Error: 0.0026 (0.26%)
Test Error: 0.0236 (2.36%)
Total Number of Support Vectors: 15



    Answer 4_d)

1) lowest training error: when C = 10000 or C = 1000000 then Training error = 0.26% 
2) lowest testing error: when C = 0.01 then Testing error = 1.65%  
   


In [1]:
!wget https://archive.ics.uci.edu/static/public/170/gisette.zip #download datafile

[0] Downloading 'https://archive.ics.uci.edu/static/public/170/gisette.zip' ...
Saving 'gisette.zip'
HTTP response 200 OK [https://archive.ics.uci.edu/static/public/170/gisette.zip]


In [6]:
!unzip gisette.zip #unzip data file

Archive:  gisette.zip
  inflating: Dataset.pdf             
   creating: GISETTE/
  inflating: GISETTE/gisette.param   
  inflating: GISETTE/gisette_test.data  
  inflating: GISETTE/gisette_train.data  
  inflating: GISETTE/gisette_train.labels  
  inflating: GISETTE/gisette_valid.data  
  inflating: gisette_valid.labels    


In [7]:
!mv ./GISETTE/* . #move to current directory

# 5] SVMs:


In [18]:
# first we need to load the data
X_train = pd.read_csv('gisette_train.data', sep='\s+', header=None)
y_train = pd.read_csv('gisette_train.labels', sep='\s+', header=None)

# now test data is not available so we use validaton as test 
X_test = pd.read_csv('gisette_valid.data', sep='\s+', header=None)
y_test = pd.read_csv('gisette_valid.labels', sep='\s+', header=None)

y_train = y_train.values.ravel()  # we need to convert dataframe to 1-d array otherwise give warning 
y_test = y_test.values.ravel()     

# now here i make function for making model we can do same in 4th question also 
def create_svm(kernel, gamma=0, degree=0, coef0=0.0):
    
    # create scm model with different para as asked in question 
    if kernel=="poly":
        svm_model = SVC(kernel=kernel, degree=degree, coef0=coef0)
    else:
        svm_model = SVC(kernel=kernel, gamma=gamma, degree=degree, coef0=coef0)
    print(f"Training the SVM model with {kernel} kernel\n")
    svm_model.fit(X_train, y_train)

    # predict on train data set
    train_predictions = svm_model.predict(X_train)
    train_error = 1 - accuracy_score(y_train, train_predictions)

    # predict on test set
    test_predictions = svm_model.predict(X_test)
    test_error = 1 - accuracy_score(y_test, test_predictions)

    # total number of support vectors
    num_support_vectors = len(svm_model.support_)

    print(f"Kernel: {kernel}")
    print(f"Train Error: {train_error:.4f} ({train_error * 100:.2f}%)")
    print(f"Test Error: {test_error:.4f} ({test_error * 100:.2f}%)")
    print(f"Number of Support Vectors: {num_support_vectors}")


    Answer 5_a)

In [19]:
print("Answer 5_a)\n")
create_svm(kernel='linear')

Answer 5_a)

Training the SVM model with linear kernel

Kernel: linear
Train Error: 0.0000 (0.00%)
Test Error: 0.0240 (2.40%)
Number of Support Vectors: 1084


    Answer 5_b)

In [20]:
print("Answer 5_b)\n")
create_svm(kernel='rbf', gamma=0.001)

Answer 5_b)

Training the SVM model with rbf kernel

Kernel: rbf
Train Error: 0.0000 (0.00%)
Test Error: 0.5000 (50.00%)
Number of Support Vectors: 6000


    Answer 5_b)

In [21]:
print("Answer 5_c)\n")
create_svm(kernel='poly',gamma=0, degree=2, coef0=1.0)

Answer 5_c)

Training the SVM model with poly kernel

Kernel: poly
Train Error: 0.0005 (0.05%)
Test Error: 0.0200 (2.00%)
Number of Support Vectors: 1332


<i><b>1. Which kernel yields the lowest training error? </i></b>   
Answer: rbf kernal because in the case polynomial training error is 0.05%  
